### PR TITLE
Support RRef.backward() for Owner RRefs.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46641 Support RRef.backward() for User RRefs.
* **#46568 Support RRef.backward() for Owner RRefs.**

This PR adds support for an RRef.backward() API. This would be useful
in applications like pipeline parallelism as described here:
https://github.com/pytorch/pytorch/issues/44827

This PR only adds support for owner RRefs, remote RRef support will be added in
a follow up PR.

Differential Revision: [D24406311](https://our.internmc.facebook.com/intern/diff/D24406311/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/init.cpp [hunk 1]
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
...
+              py::arg("retain_graph") = false,
+              py::call_guard<py::gil_scoped_release>(),
+              R"(
+                  Runs the backward pass using the RRef as the root of the
+                  backward pass. If ``dist_autograd_ctx_id`` is provided,
+                  we perform a distributed backward pass using the provided
+                  ctx_id starting from the owner of the RRef. In this case,
+                  :meth:`~torch.distributed.autograd.get_gradients` should be
+                  used to retrieve the gradients. If ``dist_autograd_ctx_id``
+                  is None, it is assumed that this is a local autograd graph
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp:413]: nit: double quote None

File: torch/csrc/distributed/rpc/init.cpp [hunk 2]
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
...
+                  and we only perform a local backward pass. The value of the
+                  RRef is expected to be a scalar Tensor.
+
+                Arguments:
+                    dist_autograd_ctx_id (int, optional): The distributed
+                        autograd context id for which we should retrieve the
+                        gradients (default: -1).
+                    retain_graph(bool, optional): If False, the graph used to
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp:421]: nit: double quote `False`

File: torch/csrc/distributed/rpc/init.cpp [hunk 3]
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
...
+                        compute the grad will be freed. Note that in nearly all
+                        cases setting this option to True is not needed and
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp:423]: ditto: double quote `True`

File: torch/csrc/distributed/rpc/init.cpp [hunk 4]
```diff
@@ -393,6 +393,43 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
...
+                        often can be worked around in a much more efficient way.
+                        Usually, you need to set this to True to run backward
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp:425]: ditto

File: torch/csrc/distributed/rpc/py_rref.cpp [hunk 5]
```diff
@@ -283,6 +285,34 @@ c10::IValue PyRRef::toIValue() const {
   return IValue(rrefPtr);
 }
 
+void PyRRef::backward(int64_t dist_autograd_ctx_id, bool retain_graph) {
+  if (rref_->isOwner()) {
+    auto value =
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/py_rref.cpp:290]: can this be `const&`?

File: torch/csrc/distributed/rpc/py_rref.cpp [hunk 6]
```diff
@@ -283,6 +285,34 @@ c10::IValue PyRRef::toIValue() const {
...
+        c10::static_intrusive_pointer_cast<const OwnerRRef>(rref_)->getValue();
+    TORCH_CHECK(
+        value.isTensor(), "RRef should contain a tensor for .backward()");
+    auto root = value.toTensor();
+    TORCH_CHECK(root.requires_grad(), "requires_grad not set on RRef's value");
+    TORCH_CHECK(
+        root.numel() == 1,
+        "RRef does not contain a scalar, all roots need to be scalar");
+    TORCH_CHECK(
+        root.grad_fn(), "RRef does not have a valid gradient function.");
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/py_rref.cpp:300]: Curious, do we need these three checks, or did the autograd engine already do that?

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 7]
```diff
@@ -3635,6 +3635,35 @@ def test_cannot_infer_backend_from_options(self):
...
             )
 
+    @dist_init
+    def test_local_rref_backward(self):
+        t = torch.rand(10, 10, requires_grad=True)
+        rref = rpc.RRef(t.sum())
+        rref.backward()
+        self.assertEqual(torch.ones_like(t), t.grad)
+
+        with dist_autograd.context() as context_id:
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:3645]: shall we add a test where part of the distributed autograd graph is on a different rank?

**Review by mrshenli — APPROVED** [review]
LGTM, added some minor comments.

File: torch/csrc/distributed/rpc/py_rref.cpp [hunk 8]
```diff
@@ -283,6 +285,34 @@ c10::IValue PyRRef::toIValue() const {
...
+      auto rootEdge = torch::autograd::impl::gradient_edge(root);
+      auto grad = at::ones_like(root, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+      torch::autograd::Engine::get_default_engine().execute(
+          {rootEdge}, {grad}, retain_graph, false);
+    } else {
+      autograd::DistEngine::getInstance().execute(
+          dist_autograd_ctx_id, {root}, retain_graph);
+    }
+  } else {
+    // TODO
```
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/py_rref.cpp:312]: Should we raise an error for now for UserRRefs or do we anticipate to land https://github.com/pytorch/pytorch/pull/46641#pullrequestreview-515149684 together/soon after this one?

File: torch/csrc/distributed/rpc/py_rref.cpp (hunk 6 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/py_rref.cpp:300]: These were unnecessary, I called the CPP APIs instead and these checks were in that API.

File: torch/csrc/distributed/rpc/py_rref.cpp (hunk 8 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/py_rref.cpp:312]: I'd anticipate to land https://github.com/pytorch/pytorch/pull/46641 soon and don't think anyone would be using this in fbcode in the mean time.

**Pull request closed without merging**